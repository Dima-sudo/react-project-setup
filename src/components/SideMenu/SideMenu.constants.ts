/* eslint-disable no-unused-vars */

export enum DataTestIds {
    COMPONENT_CONTAINER = 'sidemenu-main__container',
    TOP_CONTAINER = 'sidemenu-top__container',
    NEW_DOC_BUTTON = 'sidemenu-document__button',
    ARROW_TOGGLE_BUTTON = 'sidemenu-toggle__button',
    APP_NAME = 'sidemenu-top__appname',
}
