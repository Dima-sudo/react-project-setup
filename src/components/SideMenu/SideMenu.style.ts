import { styled } from '@mui/material/styles';

import {
    ListItem as ListItemComponent,
    ListItemButton as ListItemButtonComponent,
    ListItemText as ListItemTextComponent,
    ListItemIcon as ListItemIconComponent,
} from '@mui/material';
import SvgIconComponent from '../SvgIcon/SvgIcon';
import ButtonComponent from '../Button/Button';

export const ComponentContainer = styled('div', {
    shouldForwardProp: (prop) => prop !== 'expanded',
})<{ expanded?: boolean }>(
    ({
        expanded,
        theme: {
            custom: { colors, params },
        },
    }) => ({
        height: '100vh',
        width: expanded ? params.sidebarMenuWidth : '5.75rem',
        borderRight: `2px solid ${colors.gray30}`,
        color: colors.text1,
        display: 'flex',
        flexDirection: 'column',
        transition: params.transitionDefault,
    })
);

export const TopContainer = styled('div', {
    shouldForwardProp: (prop) => prop !== 'expanded',
})<{ expanded?: boolean }>(({ expanded }) => ({
    padding: expanded ? '2rem' : '2rem 0',
    marginBottom: '1rem',
    textAlign: 'center',
    display: 'flex',
    flexDirection: 'column',
}));

export const SvgIcon = styled(SvgIconComponent)(() => ({
    display: 'flex',
    justifyContent: 'center',
    marginBottom: '1.5rem',
}));

export const CompanyNameTitle = styled('div', {
    shouldForwardProp: (prop) => prop !== 'expanded',
})<{ expanded: boolean }>(
    ({
        expanded,
        theme: {
            custom: { params },
        },
    }) => ({
        display: 'flex',
        justifyContent: expanded ? 'space-between' : 'center',
        alignItems: 'center',
        whiteSpace: 'nowrap',
        marginBottom: '0.75rem',
        fontSize: '1rem',
        letterSpacing: '0.1rem',
        textAlign: 'left',
        transition: params.transitionDefault,
        textTransform: 'uppercase',
    })
);

export const ListItem = styled(ListItemComponent, {
    shouldForwardProp: (prop) => prop !== 'expanded',
})<{ expanded: boolean }>(({ expanded }) => ({
    display: 'flex',
    justifyContent: expanded ? 'unset' : 'center',
    '& .MuiListItemIcon-root': {
        minWidth: 'unset',
        paddingRight: expanded ? '1rem' : 0,
    },
}));

export const ListItemButton = styled(ListItemButtonComponent, {
    shouldForwardProp: (prop) =>
        prop !== 'expanded' && prop !== 'isItemSelected',
})<{
    isItemSelected?: boolean;
    expanded: boolean;
}>(
    ({
        isItemSelected,
        expanded,
        theme: {
            custom: { colors, params },
        },
    }) => ({
        paddingLeft: expanded ? '2.225rem' : '1rem',
        backgroundColor: isItemSelected ? colors.blue10 : 'unset',
        transition: params.transitionDefault,
        height: '3.5rem',
        transform: isItemSelected
            ? `${params.scaleTransition1} ${
                  expanded ? 'translateX(-0.2rem)' : 'translateX(-0.05rem)'
              }`
            : undefined,

        '&:hover': {
            backgroundColor: isItemSelected ? colors.blue10 : null,
            transition: params.transitionDefault,
        },
    })
);

export const BottomContainer = styled('div')(() => ({
    flex: 1,
    display: 'flex',
    alignItems: 'end',
    marginBottom: '1rem',
}));

export const NewDocumentButton = styled(ButtonComponent, {
    shouldForwardProp: (prop) => prop !== 'expanded',
})<{
    expanded: boolean;
}>(
    ({
        expanded,
        theme: {
            custom: { params },
        },
    }) => ({
        transition: params.transitionDefault,
        width: expanded ? '100%' : '3rem',
        height: '3rem',
        minWidth: expanded ? 'fit-content' : 'unset',
        whiteSpace: 'nowrap',
    })
);

export const MenuToggleArrow = styled(SvgIconComponent, {
    shouldForwardProp: (prop) => prop !== 'expanded',
})<{
    expanded: boolean;
}>(
    ({
        expanded,
        theme: {
            custom: { params },
        },
    }) => ({
        transition: params.transitionDefault,
        transform: expanded ? 'rotate(180deg)' : 'unset',
    })
);

export const ListItemText = styled(ListItemTextComponent)(() => ({
    whiteSpace: 'nowrap',
}));

export const ListItemIcon = styled(ListItemIconComponent)(() => ({
    display: 'flex',
    justifyContent: 'center',
}));

export const PlusIcon = styled(SvgIconComponent, {
    shouldForwardProp: (prop) => prop !== 'expanded',
})<{ expanded: boolean }>(({ expanded }) => ({
    paddingRight: expanded ? '0.25rem' : 0,
}));
