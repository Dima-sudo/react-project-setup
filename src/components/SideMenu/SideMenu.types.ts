import { IconNamesEnum } from '../SvgIcon/SvgIcon.constants';

export interface SideMenuItem {
    label: string;
    iconName?: IconNamesEnum | null;
    route: string;
}
