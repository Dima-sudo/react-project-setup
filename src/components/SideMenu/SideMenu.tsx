import React, { useState, useEffect } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import { useTranslation } from 'react-i18next';

import * as S from './SideMenu.style';

import { SideMenuItem } from './SideMenu.types';
import SvgIcon from '../SvgIcon/SvgIcon';
import { theme } from '../../theme';
import { IconNamesEnum } from '../SvgIcon/SvgIcon.constants';
import { colors } from '../../theme/colors';
import Animate from '../Animate/Animate';
import { DataTestIds } from './SideMenu.constants';

interface IProps {
    items?: SideMenuItem[];
    bottomItems?: SideMenuItem[];
    hasTopContainer?: boolean;
}

const SideMenu: React.FC<IProps> = ({
    items,
    bottomItems,
    hasTopContainer = true,
}) => {
    const navigate = useNavigate();
    const { t } = useTranslation();
    const location = useLocation();

    const [expanded, setExpanded] = useState<boolean>(true);

    const [selectedListItem, setSelectedListItem] =
        useState<SideMenuItem | null>(null);

    const onClickListItem = (item: SideMenuItem) => {
        navigate(item.route);
        setSelectedListItem(item);
    };

    useEffect(() => {
        const path = location.pathname;
        const currentItem = items?.find((item) => item.route === path);

        if (currentItem) {
            setSelectedListItem(currentItem);
        }
    }, [items, location.pathname]);

    return (
        <S.ComponentContainer
            {...{ expanded }}
            data-testid={DataTestIds.COMPONENT_CONTAINER}
        >
            {hasTopContainer && (
                <S.TopContainer
                    {...{ expanded }}
                    data-testid={DataTestIds.TOP_CONTAINER}
                >
                    <S.CompanyNameTitle
                        {...{ expanded }}
                        data-testid={DataTestIds.APP_NAME}
                    >
                        {expanded && (
                            <Animate
                                isVisible={expanded}
                                animationType="slideLeft"
                                durationSeconds={0.2}
                            >
                                {t('sidebar.appName')}
                            </Animate>
                        )}

                        <S.MenuToggleArrow
                            data-testid={DataTestIds.ARROW_TOGGLE_BUTTON}
                            iconName={IconNamesEnum.ARROW_RIGHT_LARGE}
                            color={colors.gray70}
                            onClick={() => setExpanded(!expanded)}
                            width="1.5rem"
                            height="1.5rem"
                            {...{ expanded }}
                            isButton
                        />
                    </S.CompanyNameTitle>

                    <div>
                        <S.NewDocumentButton
                            {...{ expanded }}
                            data-testid={DataTestIds.NEW_DOC_BUTTON}
                        >
                            {expanded ? (
                                <>
                                    <S.PlusIcon
                                        iconName={IconNamesEnum.PLUS}
                                        {...{ expanded }}
                                    />
                                    {t('sidebar.startButton')}{' '}
                                </>
                            ) : (
                                <S.PlusIcon
                                    iconName={IconNamesEnum.PLUS}
                                    {...{ expanded }}
                                />
                            )}
                        </S.NewDocumentButton>
                    </div>
                </S.TopContainer>
            )}

            {items && (
                <div>
                    {items.map((item) => {
                        const isItemSelected =
                            selectedListItem?.route === item.route;
                        const { colors } = theme;

                        return (
                            <S.ListItemButton
                                key={item.label}
                                isItemSelected={isItemSelected}
                                {...{ expanded }}
                                onClick={() => onClickListItem(item)}
                            >
                                <S.ListItem disablePadding {...{ expanded }}>
                                    {item.iconName && (
                                        <S.ListItemIcon>
                                            <SvgIcon
                                                iconName={item.iconName}
                                                color={
                                                    isItemSelected
                                                        ? colors.blue
                                                        : colors.gray40
                                                }
                                            />
                                        </S.ListItemIcon>
                                    )}

                                    {expanded && (
                                        <Animate
                                            isVisible={expanded}
                                            animationType="slideLeft"
                                            durationSeconds={0.2}
                                        >
                                            <S.ListItemText
                                                primary={item.label}
                                            />
                                        </Animate>
                                    )}
                                </S.ListItem>
                            </S.ListItemButton>
                        );
                    })}
                </div>
            )}

            {bottomItems && (
                <S.BottomContainer>
                    {bottomItems.map((item) => (
                        <S.ListItemButton
                            key={item.label}
                            {...{ expanded }}
                            onClick={() => onClickListItem(item)}
                        >
                            <S.ListItem disablePadding {...{ expanded }}>
                                {item.iconName && (
                                    <S.ListItemIcon>
                                        <SvgIcon
                                            iconName={item.iconName}
                                            color={theme.colors.gray40}
                                        />
                                    </S.ListItemIcon>
                                )}
                                {expanded && (
                                    <Animate
                                        isVisible={expanded}
                                        animationType="slideLeft"
                                        durationSeconds={0.2}
                                    >
                                        <S.ListItemText primary={item.label} />
                                    </Animate>
                                )}
                            </S.ListItem>
                        </S.ListItemButton>
                    ))}
                </S.BottomContainer>
            )}
        </S.ComponentContainer>
    );
};

export default SideMenu;
