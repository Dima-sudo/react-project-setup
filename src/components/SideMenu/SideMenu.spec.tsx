import '@testing-library/jest-dom';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';

import Setup from '../../tests/setup/setup';
import SideMenu from './SideMenu';
import { DataTestIds } from './SideMenu.constants';

const mockedUseNavigate = jest.fn();
const mockedUseLocation = jest.fn();

jest.mock('react-router-dom', () => ({
    // Import non-mocked library and use other functionalities and hooks
    ...jest.requireActual('react-router-dom'),

    // Mock the required hooks
    useNavigate: () => mockedUseNavigate,
    useLocation: () => mockedUseLocation,
}));

describe('Tests SideMenu component functionality', () => {
    it('Snapshots component', () => {
        render(
            <Setup>
                <SideMenu />
            </Setup>
        );

        expect(
            screen.getByTestId(DataTestIds.COMPONENT_CONTAINER)
        ).toMatchSnapshot();
    });

    it('Snapshots component', () => {
        render(
            <Setup>
                <SideMenu />
            </Setup>
        );

        expect(
            screen.getByTestId(DataTestIds.COMPONENT_CONTAINER)
        ).toMatchSnapshot();
    });

    it('Expands and collapses correctly', async () => {
        render(
            <Setup>
                <SideMenu />
            </Setup>
        );

        const arrowButton = screen.getByTestId(DataTestIds.ARROW_TOGGLE_BUTTON);

        const logo = screen.queryByTestId(DataTestIds.APP_NAME);

        expect(logo).toBeTruthy();
        expect(logo).toHaveTextContent('FW Optimization');

        fireEvent.click(arrowButton);

        await waitFor(() => {
            const collapsedLogoTextEl = screen.queryByTestId(
                DataTestIds.APP_NAME
            );
            expect(collapsedLogoTextEl).not.toHaveTextContent(
                'FW Optimization'
            );
        });
    });
});
