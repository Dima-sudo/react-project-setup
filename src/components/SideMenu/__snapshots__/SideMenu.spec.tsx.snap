// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests SideMenu component functionality Snapshots component 1`] = `
<div
  class="css-1087ewh"
  data-testid="sidemenu-main__container"
>
  <div
    class="css-tzo0h9"
    data-testid="sidemenu-top__container"
  >
    <div
      class="css-rw3fwr"
      data-testid="sidemenu-top__appname"
    >
      <div
        style="opacity: 0; transform: translateX(-20px) translateZ(0);"
      >
        FW Optimization
      </div>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-1uvutay-MuiButtonBase-root-MuiIconButton-root"
        data-testid="sidemenu-toggle__button"
        fill="#575F6E"
        height="1.5rem"
        tabindex="0"
        type="button"
        width="1.5rem"
      >
        <svg
          fill="none"
          height="31"
          viewBox="0 0 31 31"
          width="35"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            clip-path="url(#clip0_486_4805)"
          >
            <path
              d="M13.125 7.75L21.875 15.5L13.125 23.25"
              stroke="#667085"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </g>
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-5noyqf-MuiButtonBase-root-MuiButton-root"
        data-testid="sidemenu-document__button"
        tabindex="0"
        type="button"
      >
        <div
          class="css-tcyb2d"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </div>
        New Document
         
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`Tests SideMenu component functionality Snapshots component 2`] = `
<div
  class="css-1087ewh"
  data-testid="sidemenu-main__container"
>
  <div
    class="css-tzo0h9"
    data-testid="sidemenu-top__container"
  >
    <div
      class="css-rw3fwr"
      data-testid="sidemenu-top__appname"
    >
      <div
        style="opacity: 0; transform: translateX(-20px) translateZ(0);"
      >
        FW Optimization
      </div>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-1uvutay-MuiButtonBase-root-MuiIconButton-root"
        data-testid="sidemenu-toggle__button"
        fill="#575F6E"
        height="1.5rem"
        tabindex="0"
        type="button"
        width="1.5rem"
      >
        <svg
          fill="none"
          height="31"
          viewBox="0 0 31 31"
          width="35"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            clip-path="url(#clip0_486_4805)"
          >
            <path
              d="M13.125 7.75L21.875 15.5L13.125 23.25"
              stroke="#667085"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </g>
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-5noyqf-MuiButtonBase-root-MuiButton-root"
        data-testid="sidemenu-document__button"
        tabindex="0"
        type="button"
      >
        <div
          class="css-tcyb2d"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </div>
        New Document
         
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
